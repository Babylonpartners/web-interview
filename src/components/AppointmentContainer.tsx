import React from 'react'

import logo from '../logo.png'
import { API_ENDPOINT } from '../config'
import '../styles/App.scss'
import { NewAppointmentHeader } from './NewAppointmentHeader'
import { NavigationBar } from './NavigationBar'
import axios from 'axios'
import AppointmentForm from './AppointmentForm'

interface State {
  availableSlots: string[],
  availableAppTypes: string[],
  availableSlotsLoading: boolean,
  user: any,
  userLoading: boolean,

  selectedConsultantType: string,
  selectedAppointmentTime: string,
  selectedAppointmentType: string,
  appointmentNotes: string,

  selectedConsultantTypeError: string,
  selectedAppointmentTimeError: string,
  selectedAppointmentTypeError: string,
  appointmentNotesError: string,
}

export default class AppointmentContainer extends React.PureComponent<{}, State> {

  
  constructor(props:any) {
    super(props)

    this.state = {
      user: '',
      userLoading: true,
      availableSlotsLoading: true,
      availableSlots: [],
      availableAppTypes: [],

      selectedConsultantType: 'gp',
      selectedAppointmentTime: '',
      selectedAppointmentType: '',
      appointmentNotes: '',

      selectedConsultantTypeError: '',
      selectedAppointmentTimeError: '',
      selectedAppointmentTypeError: '',
      appointmentNotesError: '',
    }
  }
  

  async componentDidMount() {
      
    try {
      const slotsUrl = `${API_ENDPOINT}/availableSlots`
      const slotsResponse = await fetch(slotsUrl)
      const slotsJson = await slotsResponse.json();
      this.setState({ 
        availableSlots: slotsJson, 
        availableSlotsLoading: false 
      })
     
    } catch(e) {
      console.log(e)
    }

    try {
      const userUrl = `${API_ENDPOINT}/users/1`
      const userResponse = await fetch(userUrl)
      const userJson = await userResponse.json();
      this.setState({ 
        user: userJson, 
        userLoading: false 
      })
    } catch(e) {
      console.log(e)
    }
  }

  handleAppointmentNotesChange = (e: any) => {
    this.setState({ appointmentNotes: e.target.value})
  }

  handleAppointmentTimesChange = (e: any) => {
    this.setState({ selectedAppointmentTime: e.target.value })
  }

  handleAppointmentTypesChange = (e: any) => {
    this.setState({ selectedAppointmentType: e.target.value })
  }

  handleConsultantTypesChange = (e: any) => {
    this.setState({ selectedConsultantType: e.target.value })
  }
  
  handleValidation = () => {
    let appointmentNotesError = "";

    if (this.state.appointmentNotes === "") {
      appointmentNotesError = "Please enter a note for your consultant" 
    }

    if(appointmentNotesError) {
      this.setState({ appointmentNotesError })
      return false
    }

    return true
  }

  handleSubmit = async (e: any) => {
    e.preventDefault();
    console.log(this.state)
    const isValid = await this.handleValidation()
    if(isValid) {
      const appointmentsUrl = `${API_ENDPOINT}/appointments`
      const response = await axios.post(appointmentsUrl,
          {
            userId: this.state.user.id,
            dateTime: this.state.selectedAppointmentTime,
            notes: this.state.appointmentNotes,
            type: this.state.selectedAppointmentType,
          }
        )
      await  console.log("response: ", response)
      // clear form
      this.setState({
        selectedConsultantType: 'gp',
        selectedAppointmentTime: '',
        selectedAppointmentType: '',
        appointmentNotes: '',
      })
      console.log("this.state.appointmentNotes: ", this.state.appointmentNotes)
    }
  }
  
  
  

  render() {

    
    return (

      <>
      <NavigationBar logo={logo} />
      <div className="app">
     
     
      {/* {this.props.userLoading ?
        <div data-testid="userLoading">
          loading...
        </div>  
        : */}

        <NewAppointmentHeader 
          user={this.state.user}
        />

        {/* } */}



        {/* {this.state.availableSlotsLoading ?
        <div data-testid="availableSlotsLoading">
          loading...
        </div>  
        :
        <> */}

        <AppointmentForm 
          availableSlots={this.state.availableSlots}
          availableAppTypes={this.state.availableAppTypes}
          availableSlotsLoading={this.state.availableSlotsLoading}
          user={this.state.user}
          userLoading={this.state.userLoading}

          selectedConsultantType={this.state.selectedConsultantType}
          selectedAppointmentTime={this.state.selectedAppointmentTime}
          selectedAppointmentType={this.state.selectedAppointmentType}
          appointmentNotes={this.state.appointmentNotes}

          selectedConsultantTypeError={this.state.selectedConsultantTypeError}
          selectedAppointmentTimeError={this.state.selectedAppointmentTimeError}
          selectedAppointmentTypeError={this.state.selectedAppointmentTypeError}
          appointmentNotesError={this.state.appointmentNotesError}
        
          handleSubmit={this.handleSubmit}
          handleAppointmentNotesChange={this.handleAppointmentNotesChange}
          handleAppointmentTimesChange={this.handleAppointmentTimesChange}
          handleAppointmentTypesChange={this.handleAppointmentTypesChange}
          handleConsultantTypesChange={this.handleConsultantTypesChange}
        />
       
        {/* </>
        } */}
      </div>
      </>
    )
  }
}
