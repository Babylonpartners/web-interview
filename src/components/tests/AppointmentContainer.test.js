import React from 'react'
import { render, cleanup, waitForElement } from '@testing-library/react'
import AppointmentContainer from '../AppointmentContainer'

global.fetch = require('jest-fetch-mock')

afterEach(() => {
  cleanup()
  console.error.mockClear()
})

const user = {
  avatar: 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAIAA',
  dateOfBirth: '1971-02-03',
  firstName: 'Jane',
  lastName: 'Doe',
}

const availableSlots = [
  {
    id: 1,
    consultantType: ['gp'],
    appointmentType: ['audio', 'video'],
    time: '2019-11-27T10:11:00.000Z',
  },
  {
    id: 2,
    consultantType: ['specialist', 'gp'],
    appointmentType: ['audio', 'video'],
    time: '2019-12-01T14:16:30.000Z',
  },
]

console.error = jest.fn()

test('<AppointmentContainer>', async () => {
  const { getByTestId, debug, queryByTestId, container } = render(
    <AppointmentContainer />
  )
  expect(container.firstChild).toMatchSnapshot()

  expect(queryByTestId('navigationBarComponent')).toBeTruthy()
  expect(queryByTestId('babylonLogo')).toBeTruthy()
  expect(queryByTestId('appointmentForm')).toBeTruthy()

  expect(queryByTestId('appointmentNotesComponent')).toBeTruthy()
})

test('<AppointmentContainer> available slots fetch', async () => {
  fetch.mockResponseOnce(JSON.stringify(availableSlots))

  const { getByTestId, debug } = render(<AppointmentContainer />)

  await waitForElement(() => getByTestId('consultantTypesComponent'))
  expect(getByTestId('consultantTypesComponent')).toBeTruthy()
})

test('<AppointmentContainer> user fetch', async () => {
  fetch.mockResponseOnce(JSON.stringify(user))

  const { getByTestId } = render(<AppointmentContainer />)

  await waitForElement(() => getByTestId('userName'))
  expect(getByTestId('userName')).toBeTruthy()
})
