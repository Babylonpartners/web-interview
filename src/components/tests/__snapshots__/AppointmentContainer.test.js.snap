// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppointmentContainer> 1`] = `
<div
  class="app-header"
  data-testid="navigationBarComponent"
>
  <img
    alt="Babylon Health"
    class="app-logo"
    data-testid="babylonLogo"
    src="logo.png"
  />
</div>
`;
